### PR TITLE
[WIP] Refactored template resolution out of New3Command

### PR DESCRIPTION
**Pull request by seancpeters**

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 1]
```diff
@@ -525,25 +375,26 @@ private IReadOnlyList<ITemplateInfo> TemplatesToDisplayInfoAbout
...
                 IEnumerable<ITemplateInfo> templateList;
 
-                if (UnambiguousTemplateGroupToUse != null && UnambiguousTemplateGroupToUse.Count > 0)
+                if (_templateResolutionResult.UnambiguousTemplateGroupToUse != null && _templateResolutionResult.UnambiguousTemplateGroupToUse.Count > 0)
                 {
-                    templateList = UnambiguousTemplateGroupToUse.Select(x => x.Info);
+                    templateList = _templateResolutionResult.UnambiguousTemplateGroupToUse.Select(x => x.Info);
                 }
-                else if (!string.IsNullOrEmpty(TemplateName) && _matchedTemplatesWithSecondaryMatchInfo != null && _matchedTemplatesWithSecondaryMatchInfo.Count > 0)
+                else if (!string.IsNullOrEmpty(TemplateName) && _templateResolutionResult.MatchedTemplatesWithSecondaryMatchInfo != null && _templateResolutionResult.MatchedTemplatesWithSecondaryMatchInfo.Count > 0)
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Should we expose a property on the resolution result that is the result of the last two parts of the expression? Something like HasMatchedTemplatesWithSecondaryMatchInfo?

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 2]
```diff
@@ -525,25 +375,26 @@ private IReadOnlyList<ITemplateInfo> TemplatesToDisplayInfoAbout
...
                 {   // without template name, it's not reasonable to do secondary matching
-                    templateList = _matchedTemplatesWithSecondaryMatchInfo.Select(x => x.Info);
+                    templateList = _templateResolutionResult.MatchedTemplatesWithSecondaryMatchInfo.Select(x => x.Info);
                 }
-                else if (_matchedTemplates != null && _matchedTemplates.Any(x => x.IsMatch))
+                else if (_templateResolutionResult.CoreMatchedTemplates != null && _templateResolutionResult.CoreMatchedTemplates.Any(x => x.IsMatch))
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Same sort of "Has" bool suggestion

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 1 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: My first thought on this was "yes, absolutely!". But we sometimes store "matches" in the result lists that are only partial matches, or even have some mismatch info, e.g. invalid parameters or invalid parameter values. So I'm a bit wary that the "Has" properties could be misleading... then again the lists can be misleading too. I want to think about it a bit more, but the current situation is definitely a bit kludgy.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 3]
```diff
@@ -61,8 +59,10 @@ public New3Command(string commandName, ITemplateEngineHost host, ITelemetryLogge
             CommandName = commandName;
             _paths = new Paths(EnvironmentSettings);
             _onFirstRun = onFirstRun;
-
+            _hostDataLoader = new HostSpecificDataLoader(EnvironmentSettings.SettingsLoader);
             _commandInput = commandInput;
+
+            EnvironmentSettings.Host.TryGetHostParamDefault("prefs:language", out _defaultLanguage);
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Explicitly set _defaultLanguage to an appropriate value (null) when TryGetHostParamDefault returns false

Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: We can also reference host directly here

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 4]
```diff
@@ -17,31 +17,29 @@
...
     {
-        private static readonly IReadOnlyCollection<MatchLocation> NameFields = new HashSet<MatchLocation> { MatchLocation.Name, MatchLocation.ShortName, MatchLocation.Alias };
         private HostSpecificTemplateData _hostSpecificTemplateData;
-        private IReadOnlyList<IFilteredTemplateInfo> _matchedTemplates;
         private readonly ITelemetryLogger _telemetryLogger;
         private readonly TemplateCreator _templateCreator;
         private readonly SettingsLoader _settingsLoader;
         private readonly AliasRegistry _aliasRegistry;
         private readonly Paths _paths;
         private readonly ExtendedTemplateEngineHost _host;
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Looks like this is no longer used

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 5]
```diff
@@ -61,8 +59,10 @@ public New3Command(string commandName, ITemplateEngineHost host, ITelemetryLogge
             CommandName = commandName;
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Looks like the `host = _host = new ExtendedTemplateEngineHost(host, this);` no longer requires the `_host =` portion either

File: src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs [hunk 6]
```diff
@@ -47,38 +47,6 @@ public TemplateCache(IEngineEnvironmentSettings environmentSettings, JObject par
         [JsonProperty]
         public IReadOnlyList<TemplateInfo> TemplateInfo { get; set; }
 
-        public IReadOnlyCollection<IFilteredTemplateInfo> List(bool exactMatchesOnly, params Func<ITemplateInfo, MatchInfo?>[] filters)
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs]: Let's preserve this signature, call the new location of this method & mark it as obsolete (maybe?)

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 1 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: This turned out to be not so bad. I added the commonly used patterns as properties on the result class.

**Pull request merged**